zproject/email_backends.py

```diff
@@ -108,3 +110,46 @@ def send_messages(self, email_messages: Sequence[EmailMessage]) -> int:
                 email_log_url = settings.ROOT_DOMAIN_URI + "/emails"
                 logging.info("Emails sent in development are available at %s", email_log_url)
         return num_sent
+
+
+class PersistentSMTPEmailBackend(EmailBackend):
+    def _open(self, **kwargs: Any) -> bool | None:
+        is_opened = super().open()
+        if is_opened:
+            self.opened_at = timezone_now()
+            return True
+
+        return is_opened
+
+    @override
+    def open(self, **kwargs: Any) -> bool | None:
+        is_opened = self._open()
+        status = None
+
+        time_elapsed = (timezone_now() - self.opened_at).seconds / 60
+        if (
+            settings.EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES is None
+            or time_elapsed <= settings.EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES
+        ):
+            # No-op to ensure that we don't return a connection that has been
+            # closed by the mail server.
+            try:
+                assert self.connection is not None
+                status = self.connection.noop()[0]
+            except Exception:
+                pass
+
+        if status is None or status != 250:
+            # Close and connect again.
+            self.close()
+            is_opened = self._open()
```

Comment:
This looks like if we have `EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES = 0`:

1. We open the connection on line 126.
2. We calculate that it's been 0.00001 seconds (or whatever) since we opened it, so we don't drop into the `if`
3. `status is None` so we close the connection and re-open it

Which means we're wasting our time and the SMTP server's connecting _twice_ for every message we send.

Reply:
Fixed, this code block does not run on that setting being set to 0.